crates/chain/src/persist.rs

```diff
@@ -0,0 +1,89 @@
+use core::{convert::Infallible, marker::PhantomData};
+
+use crate::Append;
+
+/// `Persist` wraps a [`PersistBackend`] (`B`) to create a convenient staging area for changes (`C`)
+/// to the tracker (`T`) before they are persisted.
+///
+/// Not all changes to the tracker, which is an in-memory representation of wallet/blockchain
+/// data, needs to be written to disk right away, so [`Persist::stage`] can be used to *stage*
+/// changes first and then [`Persist::commit`] can be used to write changes to disk.
+#[derive(Debug)]
+pub struct Persist<B, T, C> {
+    backend: B,
+    stage: C,
+    marker: PhantomData<T>,
+}
+
+impl<B, T, C> Persist<B, T, C>
+where
+    B: PersistBackend<T, C>,
+    C: Default + Append,
+{
+    /// Create a new [`Persist`] from [`PersistBackend`].
+    pub fn new(backend: B) -> Self {
+        Self {
+            backend,
+            stage: Default::default(),
+            marker: Default::default(),
+        }
+    }
+
+    /// Stage a `changeset` to be commited later with [`commit`].
+    ///
+    /// [`commit`]: Self::commit
+    pub fn stage(&mut self, changeset: C) {
+        self.stage.append(changeset)
+    }
+
+    /// Get the changes that have not been commited yet.
+    pub fn staged(&self) -> &C {
+        &self.stage
+    }
+
+    /// Commit the staged changes to the underlying persistance backend.
+    ///
+    /// Returns a backend-defined error if this fails.
+    pub fn commit(&mut self) -> Result<(), B::WriteError> {
+        let mut temp = C::default();
+        core::mem::swap(&mut temp, &mut self.stage);
+        self.backend.write_changes(&temp)
+    }
+}
+
+/// A persistence backend for [`Persist`].
+///
+/// `T` represents the tracker, the in-memory data structure which we wish to persist.
+pub trait PersistBackend<T, C> {
```

Comment:
Shouldn't this be an [`#[async_trait]`](https://docs.rs/async-trait/latest/async_trait/)? IO operations like write / load should generally not be executed synchronously - especially if the persistence backend is a networked database.

Reply:
This is a good point. Some background on this trait. We'd rather this trait not exist but in order to preserve `Wallet`'s (over) simple API that automatically persists every change you make to it we had to add this trait. I think that if you are using a networked backend you are doing something sufficiently sophisticated to justify not using `bdk::Wallet` (and not using this trait) which is more meant as an example starting point kind of thing.

@evanlinjin thoughts?

Reply:
Yes.

In general, the structures in `bdk_chain` have a simple API and easy to design persistence around without requiring a trait. Methods that make changes return a "changeset" which the caller can persist.

Reply:
Maybe if we want to do an async version of `Wallet`, then it may be a good idea to have an async trait.

Reply:
I see so for more sophisticated use cases you suggest using `bdk_chain` directly. Thanks for this insight!